library/core/src/array/mod.rs

```diff
@@ -415,11 +415,98 @@ impl<T, const N: usize> [T; N] {
             dst.write(f(src));
             guard.initialized += 1;
         }
-        // FIXME: Convert to crate::mem::transmute once it works with generics.
-        // unsafe { crate::mem::transmute::<[MaybeUninit<U>; N], [U; N]>(dst) }
         crate::mem::forget(guard);
         // SAFETY: At this point we've properly initialized the whole array
         // and we just need to cast it to the correct type.
-        unsafe { crate::mem::transmute_copy::<_, [U; N]>(&dst) }
+        unsafe { assume_init_array(dst) }
+    }
+
+    /// Maps an `[&T; N]` or `[&mut T; N]` to a `[T; N]` by copying the contents of the array.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(array_methods)]
+    ///
+    /// let x = 12;
+    /// let ref_array = [&x, &2, &3];
+    /// let copied = ref_array.copied();
+    /// assert_eq!(copied, [12, 2, 3]);
+    /// ```
+    ///
+    /// ```
+    /// #![feature(array_methods)]
+    ///
+    /// let mut x = 12;
+    /// let ref_array = [&mut x, &mut 2, &mut 3];
+    /// let copied = ref_array.copied();
+    /// assert_eq!(copied, [12, 2, 3]);
+    ///
+    /// *ref_array[0] += 1;
+    /// let copied_again = ref_array.copied();
+    /// assert_eq!(copied_again, [13, 2, 3]);
+    /// ```
+    #[unstable(feature = "array_methods", issue = "76118")]
+    pub fn copied(&self) -> [<T as Reference>::Pointee; N]
+    where
+        T: Reference<Pointee: Copy>,
+    {
+        // No guard needed here because these are all bitwise copies, which cannot fail.
+        // (And things that are `Copy` can't be `Drop` anyway.)
+        let mut dst = MaybeUninit::uninit_array::<N>();
+        for i in 0..N {
+            dst[i].write(self[i].copy());
+        }
```

Comment:
Out of curiosity, do we get "guaranteed" bounds check elimination here? That is, am I correct that rewriting this as `dst.iter_mut().zip(self.iter())` would make this strictly worse?

Reply:
[After mir-opt](https://github.com/rust-lang/rust/blob/80fc9b0ecb29050d45b17c64af004200afd3cfc2/compiler/rustc_mir/src/transform/instcombine.rs#L64-L67) the check is against a constant, which is quite easy for LLVM to remove.  (The `for i in 0..N` loop is easily canonicalized, especially now that `Range<usize>::next` uses nowrap add.)

Reply:
Must we write byte by byte if it implements `copy`?